core/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
     public Map<Long, Task> getTasks() {
         return Collections.unmodifiableMap(new HashMap<>(tasks));
     }
+
+    /**
+     * Returns the number of currently banned tasks.
+     *
+     * Will be used in task manager stats and for debugging.
+     */
+    public int banCount() {
+        return banedParents.size();
+    }
+
+    /**
+     * Bans all tasks with the specified parent task from execution, cancels all tasks that are currently executing.
+     *
+     * This method is called when a parent task that has children is cancelled.
+     */
+    public void setBan(String parentNode, long parentId, String reason) {
+        // Set the ban first, so the newly created tasks cannot be registered
+        Tuple<String, Long> ban = new Tuple<>(parentNode, parentId);
+        synchronized (banedParents) {
+            if (lastDiscoveryNodes == null || lastDiscoveryNodes.nodeExists(parentNode)) {
+                // Only set the ban if the node is the part of the cluster
+                banedParents.put(ban, reason);
+            }
+        }
+
+        // Now go through already running tasks and cancel them
+        for (Map.Entry<Long, Task> taskEntry : tasks.entrySet()) {
+            Task task = taskEntry.getValue();
+            if (parentId == task.getParentId() && parentNode.equals(task.getParentNode())) {
+                assert task instanceof CancellableTask;
+                try {
+                    ((CancellableTask)task).cancel(reason);
+                } catch (Throwable ex) {
+                    logger.warn("Failed to cancel task with id {}", ex, task.getId());
+                }
+            }
+        }
+    }
+
+    /**
+     * Removes the ban for the specified parent task.
+     *
+     * This method is called when a previously banned task finally cancelled
+     */
+    public void removeBan(String parentNode, long parentId) {
+        banedParents.remove(new Tuple<>(parentNode, parentId));
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (lastDiscoveryNodes == null || event.nodesRemoved())  {
+            synchronized (banedParents) {
+                lastDiscoveryNodes = event.state().getNodes();
+                if (banedParents.isEmpty() == false) {
+                    // Remove all bans that were registered by nodes that are no longer in the cluster state
+                    Iterator<Tuple<String, Long>> banIterator = banedParents.keySet().iterator();
+                    while (banIterator.hasNext()) {
+                        Tuple<String, Long> nodeAndTaskId = banIterator.next();
```

Comment:
can we please unpack the tuple right away instead of using v1 v2? just easier to read